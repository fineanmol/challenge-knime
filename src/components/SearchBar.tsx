import React, { useState, useMemo } from 'react';
import { WorkflowBlock } from '../types/dataTypes';

interface SearchBarProps {
  flowBlockList: WorkflowBlock[];
  onAddBlock: (block: WorkflowBlock) => void;
}

const SearchBar: React.FC<SearchBarProps> = ({ flowBlockList, onAddBlock }) => {
  const [searchQuery, setSearchQuery] = useState<string>('');

  const addBlock = (block: WorkflowBlock) => {
    setSearchQuery('');
    onAddBlock(block);
  };

  const searchNodes = useMemo(() => {
    if (searchQuery) {
      const regExpQuery = new RegExp(searchQuery, 'i');
      return flowBlockList.filter(
        (node) => regExpQuery.test(node.name) || regExpQuery.test(node.name)
      );
    }
    return flowBlockList;
  }, [searchQuery, flowBlockList]);

  return (
    <div className="search-bar w-full bg-white flex justify-center p-4">
      <div className="relative w-full max-w-sm" data-testid="search-bar--wrapper">
        <input
          className="text-base h-10 w-full border border-gray-300 p-2 rounded focus:outline-none focus:border-blue-500"
          data-testid="search-bar--input"
          value={searchQuery}
          onChange={(e) => setSearchQuery(e.target.value)}
          type="text"
          placeholder="Add new node..."
        />
        <ul className={`absolute left-0 right-0 mt-2 bg-white border border-gray-300 rounded shadow-lg z-10 ${searchQuery ? 'block' : 'hidden'}`}>
          {searchNodes.map((node) => (
            <li
              className="flex justify-between items-center p-2 hover:bg-gray-100 cursor-pointer bg-white"
              data-testid="search-bar--list-item"
              key={node.name}
              onMouseDown={() => addBlock(node)}
            >
              <span>{node.name}</span>
              <img src={node.image} alt={node.name} className="h-10 w-10" />
            </li>
          ))}
        </ul>
      </div>
    </div>
  );
};

export default SearchBar;
