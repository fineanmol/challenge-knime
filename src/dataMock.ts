export const workflowMock = {
  nodes: [
    {
      id: 1,
      data: {
        name: "Data generator",
        description: "Source",
        image: "./assets/data_generator.jpg",
        inputs: [],
        outputs: [1, 3],
        startingPointX: 50,
        startingPointY: 50,
      },
    },
    {
      id: 2,
      data: {
        name: "CASE Switch Data (Start)",
        description: "Manipulator",
        image: "assets/case_switch.jpg",
        inputs: [2],
        outputs: [1],
        startingPointX: 250,
        startingPointY: 50,
      },
    },
    {
      id: 3,
      data: {
        name: "Cluster Assigner",
        description: "Predictor",
        image: "assets/cluster_assigner.jpg",
        inputs: [1, 3],
        outputs: [2],
        startingPointX: 450,
        startingPointY: 50,
      },
    },
  ],
  edges: [
    {
      fromNodeId: 1,
      fromInputNumber: 1,
      toNodeId: 2,
      toInputNumber: 2,
    },
    {
      fromNodeId: 2,
      fromInputNumber: 1,
      toNodeId: 3,
      toInputNumber: 3,
    },
  ],
};

export const workflowBlocksMock = [
  {
    name: "Data generator",
    description: "Source",
    image: "assets/data_generator.jpg",
    inputs: [],
    outputs: [1, 3],
  },
  {
    name: "CASE Switch Data (Start)",
    description: "Manipulator",
    image: "assets/case_switch.jpg",
    inputs: [2],
    outputs: [1],
  },
  {
    name: "Cluster Assigner",
    description: "Predictor",
    image: "assets/cluster_assigner.jpg",
    inputs: [1, 3],
    outputs: [2],
  },
];
