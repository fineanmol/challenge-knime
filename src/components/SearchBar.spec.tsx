import { render, screen, fireEvent } from '@testing-library/react';
import SearchBar from './SearchBar';
import { workflowBlocksMock } from '../dataMock';

const mockOnAddBlock = jest.fn();

describe('Render for SearchBar component', () => {
  it('renders correctly', () => {
    const { asFragment } = render(<SearchBar flowBlockList={workflowBlocksMock} onAddBlock={mockOnAddBlock} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('displays all available blocks in the options list', () => {
    render(<SearchBar flowBlockList={workflowBlocksMock} onAddBlock={mockOnAddBlock} />);

    const searchBarWrapper = screen.getByTestId('search-bar--wrapper');
    fireEvent.click(searchBarWrapper);

    const options = screen.getAllByTestId('search-bar--list-item');
    expect(options.length).toEqual(workflowBlocksMock.length);
  });

  it('only text matching search query is displayed in search input box', () => {
    render(<SearchBar flowBlockList={workflowBlocksMock} onAddBlock={mockOnAddBlock} />);

    const searchBarInput = screen.getByTestId('search-bar--input');
    fireEvent.change(searchBarInput, { target: { value: workflowBlocksMock[0].name } });

    const listItems = screen.getAllByTestId('search-bar--list-item');
    expect(listItems.length).toEqual(1);
  });

  it('calls onAddBlock function with correct data on selecting block', () => {
    render(<SearchBar flowBlockList={workflowBlocksMock} onAddBlock={mockOnAddBlock} />);

    const searchBarInput = screen.getByTestId('search-bar--input');
    fireEvent.change(searchBarInput, { target: { value: workflowBlocksMock[0].name } });

    const listItem = screen.getByText(workflowBlocksMock[0].name);
    fireEvent.mouseDown(listItem);

    expect(mockOnAddBlock).toHaveBeenCalledWith(workflowBlocksMock[0]);
  });
});
