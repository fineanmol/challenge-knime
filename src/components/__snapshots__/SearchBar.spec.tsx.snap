// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render for SearchBar component renders correctly 1`] = `
<DocumentFragment>
  <div
    class="search-bar w-full bg-white flex justify-center p-4"
  >
    <div
      class="relative w-full max-w-sm"
      data-testid="search-bar--wrapper"
    >
      <input
        class="text-base h-10 w-full border border-gray-300 p-2 rounded focus:outline-none focus:border-blue-500"
        data-testid="search-bar--input"
        placeholder="Add new node..."
        type="text"
        value=""
      />
      <ul
        class="absolute left-0 right-0 mt-2 bg-white border border-gray-300 rounded shadow-lg z-10 hidden"
      >
        <li
          class="flex justify-between items-center p-2 hover:bg-gray-100 cursor-pointer bg-white"
          data-testid="search-bar--list-item"
        >
          <span>
            Data generator
          </span>
          <img
            alt="Data generator"
            class="h-10 w-10"
            src="assets/data_generator.jpg"
          />
        </li>
        <li
          class="flex justify-between items-center p-2 hover:bg-gray-100 cursor-pointer bg-white"
          data-testid="search-bar--list-item"
        >
          <span>
            CASE Switch Data (Start)
          </span>
          <img
            alt="CASE Switch Data (Start)"
            class="h-10 w-10"
            src="assets/case_switch.jpg"
          />
        </li>
        <li
          class="flex justify-between items-center p-2 hover:bg-gray-100 cursor-pointer bg-white"
          data-testid="search-bar--list-item"
        >
          <span>
            Cluster Assigner
          </span>
          <img
            alt="Cluster Assigner"
            class="h-10 w-10"
            src="assets/cluster_assigner.jpg"
          />
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;

exports[`SearchBar component renders correctly 1`] = `
<DocumentFragment>
  <div
    class="search-bar w-full bg-white flex justify-center p-4"
  >
    <div
      class="relative w-full max-w-sm"
      data-testid="search-bar--wrapper"
    >
      <input
        class="text-base h-10 w-full border border-gray-300 p-2 rounded focus:outline-none focus:border-blue-500"
        data-testid="search-bar--input"
        placeholder="Add new node..."
        type="text"
        value=""
      />
      <ul
        class="absolute left-0 right-0 mt-2 bg-white border border-gray-300 rounded shadow-lg z-10 hidden"
      >
        <li
          class="flex justify-between items-center p-2 hover:bg-gray-100 cursor-pointer bg-white"
          data-testid="search-bar--list-item"
        >
          <span>
            Data generator
          </span>
          <img
            alt="Data generator"
            class="h-10 w-10"
            src="assets/data_generator.jpg"
          />
        </li>
        <li
          class="flex justify-between items-center p-2 hover:bg-gray-100 cursor-pointer bg-white"
          data-testid="search-bar--list-item"
        >
          <span>
            CASE Switch Data (Start)
          </span>
          <img
            alt="CASE Switch Data (Start)"
            class="h-10 w-10"
            src="assets/case_switch.jpg"
          />
        </li>
        <li
          class="flex justify-between items-center p-2 hover:bg-gray-100 cursor-pointer bg-white"
          data-testid="search-bar--list-item"
        >
          <span>
            Cluster Assigner
          </span>
          <img
            alt="Cluster Assigner"
            class="h-10 w-10"
            src="assets/cluster_assigner.jpg"
          />
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;
